### ISSUE TITLE
Excessive memory usage when printing multi-file Dataset

### ISSUE THREAD

**Issue by hadfieldnz** [pydata/xarray#1481]
I have a dataset comprising 25 output files from the ROMS ocean model. They are netCDF files ("averages" files in ROMS jargon) containing a number of variables, but most of the storage is devoted to a few time-varying oceanographic variables, either 2D or 3D in space. I have post-processed the files by packing the oceanographic variables to int32 form using the netCDF add_offset and scale_factor attributes. Each file has 100 records in the unlimited dimension (ocean_time) so the complete dataset has 2500 records. The 25 files total 56.8 GiB so would expand to roughly 230 GiB in float64 form.

I open the 25 files with xarray.open_mfdataset, concatenating along the unlimited dimension. This takes a few seconds. I then print() the resulting xarray.Dataset. This takes a few seconds more. All good so far.

But when I vary the number of these files, _n_, that I include in my xarray.Dataset I get surprising and inconvenient results. All works as expected in reasonable time with _n_ <= 8 and with _n_ >= 19. But with 9 <= _n_ <= 18, the interpreter that's processing the code (pythonw.exe via Ipython) consumes steadily more memory until the 12-14 GiB that's available on my machine is exhausted.

The attached script exposes the problem. In this case the file sequence consists of one file name repeated _n_ times. The value of _n_ currently hard-coded into the script is 10. With this value, the final statement in the script--printing the dataset--will exhaust the memory on my PC in about 10 seconds, if I fail to kill the process first.

I have put a copy of the ROMS output file here: 

ftp://ftp.niwa.co.nz/incoming/hadfield/roms_avg_0001.nc

[mgh_example_test_mfdataset.py.txt](https://github.com/pydata/xarray/files/1158093/mgh_example_test_mfdataset.py.txt):
```python
"""Explore a performance/memory bug relating to multi-file datasets
"""

import xarray as xr
import os

#%% Specify the list of files. Repeat the same file to get a mult-file dataset

root = os.path.join(
        'D:\\', 'Mirror', 'hpcf', 'working', 'hadfield',
        'work', 'cook', 'roms', 'sim34', 'run',
        'bran-2009-2012-wrfnz-1.20')
file = [os.path.join(root, 'roms_avg_0001.nc')]

file = file*10

print('Number of files:', len(file))

#%% Create a multi-file dataset with the open_mfdataset function.

ds = xr.open_mfdataset(file, concat_dim='ocean_time')

print('The dataset has been successfully opened')

#%% Print a summary

print(ds)

print('The dataset has been printed')
```

**Comment by rabernat** [comment]
Hi @hadfieldnz -- I believe this issue could be related to #1396, which was fixed in dask/dask#2364.

Could you let us know what versions of xarray and dask you are using?

```python
import xarray
import dask
print(xarray.__version__)
print(dask.__version__)
```

**Comment by hadfieldnz** [comment]
xarray 0.9.6
dask 0.14.3

**Comment by fmaussion** [comment]
0.14.3 pre-dates the fix https://github.com/dask/dask/pull/2364 mentioned above: can you try to update dask?

**Comment by hadfieldnz** [comment]
I ran "conda update dask", which upgraded me from 0.14.3 to 0.15.0.

Short report: No this has not eliminated the problem.

Long report: Today (Friday) I am on my home machine, which has only 6 GiB RAM. I confirmed earlier today with dask 0.14.3 that I can open and print the dataset with 25 files. And with 10 files IPython halts with a memory error reporting that 85% of the memory is being used. After the upgrade to 0.15.0, running the test script with 10 files, it exhausted all the RAM on my machine and locked it up within a few seconds. I will not be able to investigate this further until I get back on my work machine on Monday.

**Comment by rabernat** [comment]
Can you try calling `open_mfdataset` with the `decode_df=False` option?

**Comment by shoyer** [comment]
Our [formatting logic](https://github.com/pydata/xarray/blob/5d245b22e9500a7eb805193ba5c65bb5474a5ae1/xarray/core/formatting.py#L176) pulls out the first few values of arrays to print them in the repr. It appears that this is failing spectacularly in this case, though I'm not sure why.

Can you share a quick preview of what a single one of your constituent netCDF files looks like?

More broadly: maybe we should disable automatically printing a preview of the contents of `xarray.Dataset` objects when they have lazily loaded data in the form of dask arrays. This is convenient for interactive use in many cases (when it can be done cheaply!) but fails in many edge cases.

**Comment by jhamman** [comment]
@hadfieldnz - I think this was just fixed in #1532. Keep an eye out for the 0.10 release. Feel free to reopen if you feel there's more to do here.
